group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -731,6 +752,36 @@ private void throwIfConsumerGroupIsFull(
         }
     }
 
+    private void throwIfStaticMemberValidationFails(
+        String groupId,
+        String instanceId,
+        ConsumerGroupMember existingStaticMember,
+        int memberEpoch,
+        String memberId
+    ) {
+        if (memberEpoch != 0) {
+            // The member joined with a non-zero epoch but we haven't registered this static member
+            // This could be an unknown member for the coordinator.
+            if (existingStaticMember == null) {
+                throw Errors.UNKNOWN_MEMBER_ID.exception("No existing static member found against instance id: " +  instanceId);
+            }
+            // There already exists a different member id for the same instance id.
+            if (!existingStaticMember.memberId().equals(memberId)) {
+                log.info("Request memberId={} for static member with groupInstanceId={} " +
+                        "is fenced by existing memberId={}",
+                    memberId, instanceId, existingStaticMember.memberId());
+                throw Errors.FENCED_INSTANCE_ID.exception();
```

Comment:
Could we add something to the exception?

Reply:
Added.